torch/_functorch/_aot_autograd/runtime_wrappers.py

```diff
@@ -405,49 +402,126 @@ def runtime_wrapper(args: List[Any]):
+                runtime_args.extend([seed, offset])
+                out = compiled_fn(runtime_args)
+                out = self._functionalized_rng_runtime_epilogue(
+                    fw_metadata,
+                    out,
+                    # TODO: this won't be right for the backward when we convert the call_compiled_backward to use the wrapper
+                    fw_metadata.num_forward_returns,
+                )
+                return out
+            return compiled_fn(runtime_args)
+
+        return wrapper
+
+    # Calling convention: If we are running functionalized RNG, then outs consists
+    # of (user_outs, rng_offset)
+    def _functionalized_rng_runtime_epilogue(
+        self,
+        metadata: ViewAndMutationMeta,
+        outs,
+        offset_index,
+    ):
+        if metadata.is_rng_op_functionalized:
+            assert metadata.num_outputs_rng_offset == 1
+            new_rng_offset = outs[offset_index]
+            CUDARngStateHelper.set_new_offset(new_rng_offset)
+            if self.return_new_outs:
+                user_outs = outs[:offset_index] + outs[offset_index + 1 :]
+                return user_outs
+            else:
+                return outs
+
+        return outs
+
+
+@dataclass
+class FakifiedOutWrapper(CompilerWrapper):
+    fakified_out: Optional[Any] = None
+
+    def post_compile(
+        self,
+        compiled_fn,
+        aot_config: AOTConfig,
+        *,
+        fw_metadata: ViewAndMutationMeta,
+    ):
+        fakified_out = self.fakified_out
+
+        @wraps(compiled_fn)
+        def wrapper(runtime_args):
+            nonlocal fakified_out
+            if fakified_out is not None:
+                out = fakified_out
+                fakified_out = None
+                return out
+            return compiled_fn(runtime_args)
+
+        return wrapper
 
 
 # This wrapper handles the AOTDispatch runtime logic for tensor subclasses.
 # At runtime, we have a compiled function that knows how to operate on the domain of DenseTensor -> DenseTensor,
 # But the user might have passed us some tensor subclass inputs (or expect some subclass tensor outputs).
 # This function handles the wrapping and unwrapping of tensor subclasses at runtime.
-def aot_dispatch_subclass_wrapper(
-    runtime_fn: Callable,
-    *,
-    subclass_metas: List[Union[int, SubclassCreationMeta]],
-    num_fw_outs_saved_for_bw: Optional[int],
-) -> Callable:
-    def inner_fn(args: List[Any]):
-        unwrapped_args = unwrap_tensor_subclasses(args, is_joint_structure=False)
-        args.clear()
-        # expectation: runtime_fn is a boxed fn
-        unwrapped_outs = runtime_fn(unwrapped_args)
-        wrapped_outs = wrap_tensor_subclasses(
-            unwrapped_outs,
-            subclass_metas=subclass_metas,
-            num_fw_outs_saved_for_bw=num_fw_outs_saved_for_bw,
-            is_runtime=True,
-        )
-        return wrapped_outs
+@dataclass
+class AOTDispatchSubclassWrapper(CompilerWrapper):
```

Comment:
The subclass wrapper also has a `pre_compile` logic, which is effectively what `aot_dispatch_subclass` does: https://github.com/pytorch/pytorch/blob/main/torch/_functorch/_aot_autograd/dispatch_and_compile_graph.py#L85

Are you thinking of making that the actual `pre_compile` for this class?

Reply:
Thanks for the link, will take a look!

Reply:
Implemented a simple pre_compile for the wrapper, but in an effort not to make the current code more confusing for no reason, not going to add the call to it in aot_dispatch_*_graph just yet, until we've refactored the wrappers above it. 

The intended use of the wrapper is something like
```
subclass_wrapper = AOTDispatchSubclassWrapper(fw_only=flat_fn, trace_joint=trace_joint)
(flat_fn, flat_args, aot_config, fw_metadata) = subclass_wrapper.pre_compile(flat_fn, flat_args, aot_config, fw_metadata) # Sets maybe_subclass_meta here
... # Other precompile steps
compiled_fn = compiler_fn(flat_fn, flat_args, aot_config, fw_metadata)
... # Other postcompile steps
compiled_fn = subclass_wrapper.post_compile(compiled_fn, ...) # Uses maybe_subclass_meta here
```
But we can't do that until functionalized_fn, etc are also refactored. 

Instead I added a todo at the aot_dispatch_subclass callsites.